File: Graph-theory/207-canFinish/go/207.go
package main

import "fmt"

// BFS
func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := make([][]int, numCourses)
	for i := range numCourses {
		graph[i] = make([]int, 0)
	}
	indegree := make([]int, numCourses) // 某节点的入度
	queue := make([]int, 0)
	// 初始化图和入度
	for _, pair := range prerequisites {
		graph[pair[1]] = append(graph[pair[1]], pair[0])
		indegree[pair[0]]++
	}
	// 初始化后入度为 0 的节点就是不需要先修课的节点
	for i := range indegree {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	// 每次取出队头，将队头节点指向的其他节点入度减一
	// 相当于“把这门先修课修完了”
	studied := 0
	for len(queue) != 0 {
		u := queue[0]
		queue = queue[1:]
		studied++
		for _, v := range graph[u] {
			indegree[v]--
			if indegree[v] == 0 {
				queue = append(queue, v)
			}
		}
	}
	return studied == numCourses
}

func main() {
	prerequisites := [][]int{{1, 0}, {0, 1}}
	fmt.Println(canFinish(2, prerequisites))
}

File: Graph-theory/207-canFinish/207.cpp
#include <cstddef>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 private:
  vector<vector<int>> prerequistes_;
  unordered_map<int, vector<int>> prerequisite_map_;

 public:
  void build_prerequire_map() {
    for (const auto& pair : prerequistes_) {
      prerequisite_map_[pair[0]].push_back(pair[1]);
    }
  }

  bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    prerequistes_ = prerequisites;
    build_prerequire_map();

    unordered_set<int> to_be_studied;
    unordered_set<int> have_studied;
    for (int i = 0; i < numCourses; ++i) {
      if (prerequisite_map_[i].empty()) {
        have_studied.insert(i);
      } else {
        to_be_studied.insert(i);
      }
    }
    if (have_studied.empty()) {
      return false;
    }

    size_t last_studied_cnt = have_studied.size();
    while (true) {
      for (auto it = to_be_studied.begin(); it != to_be_studied.end();) {
        // 检查目前 course[i] 能不能学
        bool can_study = true;
        for (int required : prerequisite_map_[*it]) {
          if (!have_studied.count(required)) {
            can_study = false;
            break;
          }
        }
        if (can_study) {
          have_studied.insert(*it);
          it = to_be_studied.erase(it);
        } else {
          ++it;
        }
      }
      // 检查这轮过后，能学的课程是否有增加
      if (have_studied.size() == last_studied_cnt) {
        break;
      }
      last_studied_cnt = have_studied.size();
    }

    return int(have_studied.size()) == numCourses;
  }
};

int main() {
  vector<vector<int>> prerequisites = {{1, 0}, {0, 1}};
  Solution obj;
  cout << obj.canFinish(2, prerequisites) << endl;

  return 0;
}